File: StackUsingQueues/StackUsingQueues.js
class StackUsingQueues {
    constructor() {
        this.q1 = [];
        this.q2 = [];
    }

    push(value) {
        this.q1.push(value);
    }

    pop() {
        // Move all elements except the last one from q1 to q2
        while (this.q1.length > 1) {
            this.q2.push(this.q1.shift());
        }

        // Get the last element (top of the stack)
        const topElement = this.q1.shift();

        // Swap the queues
        [this.q1, this.q2] = [this.q2, this.q1];

        return topElement;
    }

    top() {
        // Similar to pop but re-adds the top element back to the queue
        while (this.q1.length > 1) {
            this.q2.push(this.q1.shift());
        }

        const topElement = this.q1.shift();
        this.q2.push(topElement);

        [this.q1, this.q2] = [this.q2, this.q1];

        return topElement;
    }

    isEmpty() {
        return this.q1.length === 0;
    }
}

// Example Usage
const stack = new StackUsingQueues();
stack.push(10);
stack.push(20);
console.log("Top element:", stack.top());  // Output: 20
console.log("Popped element:", stack.pop());  // Output: 20
console.log("Is stack empty?", stack.isEmpty());  // Output: false

File: StackUsingQueues/StackUsingQueues.py
from queue import Queue

class StackUsingQueues:
    def __init__(self):
        self.q1 = Queue()
        self.q2 = Queue()

    def push(self, value):
        print(f"Pushed {value}")
        self.q1.put(value)

    def pop(self):
        while self.q1.qsize() > 1:
            self.q2.put(self.q1.get())
        top_element = self.q1.get()
        print(f"Popped element: {top_element}")
        self.q1, self.q2 = self.q2, self.q1
        return top_element

    def top(self):
        while self.q1.qsize() > 1:
            self.q2.put(self.q1.get())
        top_element = self.q1.get()
        self.q2.put(top_element)
        self.q1, self.q2 = self.q2, self.q1
        print(f"Top element: {top_element}")
        return top_element

    def is_empty(self):
        empty = self.q1.empty()
        print(f"Is stack empty? {empty}")
        return empty

def run_stack_example(stack):
    stack.push(10)
    stack.push(20)
    stack.top()         # Should print: Top element: 20
    stack.pop()         # Should print: Popped element: 20
    stack.is_empty()    # Should print: Is stack empty? False
    stack.pop()         # Should print: Popped element: 10
    stack.is_empty()    # Should print: Is stack empty? True

# Running the example
stack = StackUsingQueues()
run_stack_example(stack)
